share/man/man1/proxy_send.1
.TH "Proxy_send - send proxy authentication" "1" 

.SH "Description"
.IP \(em 4
This script is automatically called by prady_runcom's profile scripts when network is inferred as \fIoffice\fP.
.IP \(em 4
We try to open https://www.duckduckgo.com/ \fBat\fP \fIduckduckgo\fP and send proxy authentication secret.
.IP \(em 4
The proxy server generally 'remembers' that our machine was authenticated and
does not demand proxy authentication again.

.SH "Setup"
.PP
Proxy secret may be provided through any of the tools explained in following sections.

.IP \(em 4
If information is available at both locations, it is loaded in order.
.IP \(em 4
Information available from the latter supersedes the former.

.SS "Password store"
.IP \(em 4
https://www.passwordstore.org \fBat\fP \fIPassword manager\fP may hold the proxy-secret.
.IP \(em 4
Pass-name to the secret \fBMUST\fP be stored in an environment variable \fI${proxy_auth}\fP.
.IP \(em 4
Secret MUST \fBNOT\fP be url-encoded.

.IP \(em 4
\fIpass-name\fP can be declared by adding following lines to
\fI${XDG_CONFIG_HOME:\-${HOME/.config}/(pvt|local).d/.<some_name>rc\fP
.IP \(em 4
setup.org \fBat\fP \fIPersonalization\fP

.RS
.nf
\fCproxy_auth="office/proxy_server.example.com/best_employee"
export proxy_auth
\fP
.fi
.RE

.IP \(em 4
Secret \fBMUST\fP be stored in either of following formats.

.RS
.TP
.ft I
plain password
.ft
.PP
Password store stores just the <password>.
<password> may be stored in password manager using a command such as:
.RS
.nf
\fCpass insert office/proxy_server.example.com/best_employee
\fP
.fi
.RE
.RE

.TP
.ft I
multi-field
.ft
.PP
Information about \fIany\fP or \fIall\fP of following fields may be provided.
.IP \(em 4
Order is not strict.

.RS
.nf
\fCprotocol: <proxy protocol>
username: <username in plain text>
password: <password in plain text>
address: <proxy server address>
port: <proxy port>

\fP
.fi
.RE

.PP
Fields may be stored in password manager using a command such as:
.RS
.nf
\fCpass insert -m office/proxy_server.example.com/best_employee
\fP
.fi
.RE
.RE

.SS "Environment declaration"
.IP \(em 4
A \fI${proxy_protocol}\fP environment variable \fBmay\fP be declared.
It can be declared by adding following lines to
\fI${XDG_CONFIG_HOME:\-${HOME/.config}/(pvt|local).d/.<some_name>rc\fP

.RS
.nf
\fCproxy_protocol="http"
export proxy_protocol

\fP
.fi
.RE

.IP \(em 4
Proxy information extraction attempt is made from \fI${proxy_protocol}_proxy\fP environment variable
.IP \(em 4
e.g. \fIhttp_proxy\fP, \fIsocks_proxy\fP, etc.
.IP \(em 4
If \fIproxy_protocol\fP remains unset, extraction is attempted from \fIall_proxy\fP and \fIhttp\fP is used as default.

.IP \(em 4
\fI${<proxy_protocol>_proxy}\fP should be declared in
\fI${XDG_CONFIG_HOME:\-${HOME/.config}/(pvt|local).d/.<some_name>rc\fP.
.IP \(em 4
Special characters (symbols) in username and password \fBMUST\fP be https://en.wikipedia.org/wiki/Percent-encoding \fBat\fP \fIurl-encoded\fP.
.IP \(em 4
Its value must be of the form:
.RS
.nf
<p_type>_proxy="<protocol>://[<encoded_username>[:<encoded_password>]]@<address>[:<port>]/"

.fi
.RE

.IP \(em 4
example for \fIall_proxy\fP
.RS
.nf
\fCall_proxy="http://best_employee:secure%21secret@192.168.1.1:1080/"
export all_proxy
\fP
.fi
.RE

Storing password in environment is strongly \fBdiscouraged\fP.

.SH "Code"
.PP
This section tangles the code (POSIX)
.SS "Check availability"
.RS
.nf
\fCaffirm_availability () {
    for _import in curl printenv grep tr; do
        if ! command -v "${_import}" >/dev/null 2>&1; then
            unset _import
            clean_exit 127 "${_import} is not installed\n"
        fi
    done
    unset _import
}

\fP
.fi
.RE

.SS "Handle global variables and exits"
.RS
.nf
\fCset_vars () {
    show=false
    pass_keys="protocol username password host port"
    proxy_header=
    if [ -z "${proxy_protocol}" ]; then
        proxy_protocol='all'
    fi
    proxy_username=
    proxy_password=
    proxy_host=
    proxy_port=
    usage="
    usage: ${0} -h
    usage: ${0} --help
    usage: ${0} [Optional Arguments*] INSTANCE
"
    help_msg="${usage}

    DESCRIPTION: |
      Auto-send proxy authentication


    Optional Arguments: |
      -h\t\t\tprint usage message and exit
      --help\t\t\tprint this help message and exit
      -s|--show\tdisplay what will be sent as header, don't send

"
}

unset_vars() {
    unset help_msg
    unset usage
    unset show
    unset proxy_port
    unset proxy_host
    unset proxy_password
    unset proxy_username
    unset pass_keys
    unset proxy_header
}


clean_exit() {
    unset_vars
    if [ -n "${1}" ] && [ "${1}" -ne "0" ]; then
        if [ -n "${2}" ]; then
            # shellcheck disable=SC2059
            printf "${2}\n" >&2
        fi
        # shellcheck disable=SC2086
        exit ${1}
    fi
    if [ -n "${2}" ]; then
        # shellcheck disable=SC2059
        printf "${2}\n"
    fi
    exit 0
}

\fP
.fi
.RE

.SS "Parse command"
.RS
.nf
\fCcli () {
    while [ $# -gt 0 ]; do
        case "${1}" in
            -h)
                # shellcheck disable=SC2059
                clean_exit 0 "${usage}"
                ;;
            --help)
                # shellcheck disable=SC2059
                clean_exit 0 "${help_msg}"
                ;;
            -s|--show)
                show=true
                shift
                ;;
            *)
                clean_exit 1 "${usage}"
        esac
    done
}

\fP
.fi
.RE

.SS "Extract proxy information from environment variable"
.RS
.nf
\fCextract_env () {
    url="${1}"
    # keep consuming URL like $@ is consumed from command line
    _proto="$(printf "%s" "${url}" | grep :// | sed -e 's,^\(.*\)://.*,\1,g')"
    url="${url#${_proto}://}"  # - protocol
    userpass="$(printf "%s" "${url}" | grep @ | cut -d@ -f1)"
    _user="${userpass%:*}"
    _pass="$(printf "%s" "${userpass}" | grep : | sed -e 's,^.*\?:\(.*\),\1,g')"
    url="$(printf "%s" "${url##${userpass}@}" | cut -d/ -f1)"  # - credentials
    _host="${url%:*}"
    _port="$(printf "%s" "${url}" | \grep '[0-9]' | sed -e 's,^.*:\([0-9]\+\)$,\1,')"
    if [ -n "${_proto}" ]; then
        proxy_protocol="${_proto}"
    fi

    if [ -n "${_user}" ]; then
        proxy_username="${_user}"
    fi

    if [ -n "${_pass}" ]; then
        proxy_password="${_pass}"
    fi

    if [ -n "${_host}" ]; then
        proxy_host="${_host}"
    fi

    if [ -n "${_port}" ]; then
        proxy_port="${_port}"
    fi

     unset _port _host _pass _user _proto userpass url
}

get_env_proxy () {
    # Parse environment variable.
    proxy_str="$(printenv "${proxy_protocol}_proxy")"
    if [ -z "${proxy_str}" ]; then
        return
    fi
    extract_env "${proxy_str}"
    unset proxy_str
}

\fP
.fi
.RE

.SS "Compile proxy information"
.RS
.nf
\fCbuild () {
    all_proxy="$(${RUNCOMDIR:-${HOME}/.runcom}/bin/proxy_extract.sh)"
    extract_env "${all_proxy}"
    get_env_proxy
}

compile_proxy () {
    if [ -z "${proxy_host}" ]; then
        return
    fi
    if [ -n "${proxy_protocol}" ]; then
        if [ "${proxy_protocol}" = "all" ]; then
            proxy_header="http://"
        else
            proxy_header="${proxy_protocol}://"
        fi
    fi
    if [ -n "${proxy_username}" ]; then
        scrt="${proxy_username}"
        if [ -n "${proxy_password}" ]; then
            scrt="${proxy_username}:${proxy_password}"
        fi
        proxy_header="${proxy_header}${scrt}@"
    fi
    proxy_header="${proxy_header}${proxy_host}"
    if [ -n "${proxy_port}" ]; then
        proxy_header="${proxy_header}:${proxy_port}"
    fi
    proxy_header="${proxy_header}/"
    unset scrt
}

\fP
.fi
.RE

.SS "Html handling"
.RS
.nf
\fCquote () {
    printf "%s" "$1" \
        | tr -d '\n' \
        | curl -Gso /dev/null -w "%{url_effective}" --data-urlencode @- "" \
        | cut -c 3-
}

send_request () {
    curl -sLf -x "${proxy_header}" "https://www.duckduckgo.com/" >/dev/null 2>&1
    case $? in
        0)
            clean_exit
            ;;
        6)
            # Couldn't resolve
            clean_exit 6
            ;;
        7)
            # No route to proxy_host
            clean_exit
            ;;
        *)
            # other error
            clean_exit "$?"
            ;;
    esac
}

\fP
.fi
.RE

.SS "Main routine call"
.RS
.nf
\fCmain() {
    # Main routine call
    affirm_availability
    set_vars
    cli "$@"
    build
    compile_proxy
    if $show; then
        printf "auth: '%s'\n" "${proxy_header}"
        clean_exit
    fi
    send_request
    clean_exit
}


main "$@"
\fP
.fi
.RE


proxy_send.py
#!/usr/bin/env python3
# -*- coding:utf-8; mode:python -*-
#
# Copyright 2020, 2021 Pradyumna Paranjape
#
# Check for network connectivity at the beginning
# This file is part of Prady_runcom.
#
# Prady_runcom is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Prady_runcom is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Prady_runcom.  If not, see <https://www.gnu.org/licenses/>.
'''
An (unsynced) secrets file should export secrets:

proxy_type,
USERNAME,
password_plain/password_html,
proxy_addr,
proxy_port

... to env.

We try to open "https://www.duckduckgo.com/" and send a proxy auth
to the server.

Proxy server generally 'remembers' that our machine was authenticated
and does not demand proxy auth again.
'''

# imports
import os
import sys
import typing
from urllib.parse import quote

from requests import get


class Proxy():
    '''
    proxy class
    '''
    def __init__(self, **kwargs) -> None:
        '''
        p_type: proxy_type: socks, http
        uname: proxy user name
        password: proxy password (url encoded)
        address: proxy address (may be ip address)
        port: proxy port

        initiate an empty proxy
        '''
        self.p_type: typing.Optional[str] = None
        self.auth: typing.Dict[str, typing.Optional[str]] = {
            'uname': None,
            'pass': None
        }
        self.target: typing.Dict[str, typing.Optional[str]] = {
            'address': None,
            'port': None
        }
        for key, val in kwargs.items():
            if key in self.auth:
                self.auth[key] = val
            elif key in self.target:
                self.target[key] = val
            else:
                raise KeyError(f"Bad keyword argument {key}")

    def __str__(self) -> str:
        '''
        string of the form
        "p_type://[USERNAME[:passwordurl]@]address[:port]/"
        '''
        if self.p_type is None:
            return ''
        if self.target['address'] is None:
            return ''
        proxy = []
        proxy.append(f"{self.p_type}://")
        if self.auth['uname'] is not None:
            proxy.append(self.auth['uname'])
            if self.auth['password'] is not None:
                proxy.append(":" + self.auth['password'])
            proxy.append("@")
        proxy.append(self.target['address'])
        if self.target['port'] is not None:
            proxy.append(":" + str(self.target['port']))
        proxy.append("/")
        return ''.join(proxy)

    def add_uname(self, uname: str, encoded: bool = False) -> None:
        '''
        uname: user name
        encoded: USERNAME is encoded?
        '''
        self.auth['uname'] = uname if encoded else quote(uname)

    def add_password(self, password: str, encoded: bool = False) -> None:
        '''
        password: proxy password
        encoded: url encoded?

        add a password
        '''
        self.auth['password'] = password if encoded else quote(password)


def build():
    '''
    build proxy dictionary from cli
    '''
    if 'http_proxy' in os.environ:
        return os.environ['http_proxy']
    # Definitions
    proxy = Proxy()
    if 'proxy_type' in os.environ:
        proxy.p_type = os.environ['proxy_type']
    else:
        print("proxy type not supplied")
        return ''
    if 'passwordplain' in os.environ:
        proxy.add_password(os.environ['passwordplain'])
    elif 'passwordhtml' in os.environ:
        proxy.add_password(os.environ['passwordhtml'], encoded=True)
    if 'USERNAME' in os.environ:
        proxy.add_uname(os.environ['USERNAME'])
    if 'proxy_addr' in os.environ:
        proxy.target['address'] = os.environ['proxy_addr']
    if 'proxy_port' in os.environ:
        proxy.target['port'] = os.environ['proxy_port']
    return {proxy.p_type: str(proxy)}


def main():
    '''
    main subroutine
    '''
    proxies = build()
    try:
        get("https://www.duckduckgo.com/", proxies=proxies)
    except ConnectionRefusedError:
        return 127
    except OSError as err:
        if 'No route to host' in str(err):
            return 0
        else:
            print(err)
            return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
